### ISSUE TITLE
Problems with channel_layer.group_send() from a Celery @task

### ISSUE THREAD

**Issue by EugenSusurrus** [CJWorkbench/channels_rabbitmq#37]
Hi,

many thanks for providing the channels-rabbitmq implementation. Since my project has a requirement for RabbitMQ and it is using celery I find it very useful to have the option of not using channels-reds.

However since a couple of days I'm struggling to send a message from a celery task to a Channels WebsocketConsumer over the channel layer. I've looked through the issues already submitted here and [this one](https://github.com/CJWorkbench/channels_rabbitmq/issues/34) seems to describe closer what I want to achieve.

in my channels `consumers.py` file I have the following:
```python
import json

from asgiref.sync import async_to_sync
from channels.generic.websocket import WebsocketConsumer

class WebsocketConnectionConsumer(WebsocketConsumer):
    """Consumer to handle websocket connections."""

    def task_assign(self, event):
        """Handles the `task.assign` event received on the channel layer."""
        async_to_sync(self.send(json.dumps(
              {
                  "event": "task.received",
                  "payload": {
                      "host_id": "id_placeholder",
                      "task_id": "id_placeholder",
                      "task": "id_placeholder"
                  }
              }
          )))

```

which I expect to be triggered by the call in the celery task:

```python
import asyncio
import json
from asgiref.sync import async_to_sync, sync_to_async
from celery import shared_task
from channels.layers import get_channel_layer

@shared_task
def pick_host_for_task_execution():
    """Celery task to pick a `Host` instance and assign it a task to execute."""

    async def send_assign_task(group_name, task, host):
        await get_channel_layer().group_send(
              group_name,
              {
                  "type": "task.assign",
                  "task_id": "id_placeholder",
                  "host_id": "id_placeholder"
              }
          )

    asyncio.run(send_assign_task(group_name="group_name", task="task_id_placeholder", host="host_id_placeholder"))

```

When I run the server and connect to it using websockets, the connected client doesn't seem to get the message from the websocket consumer, which should be triggered by the celery task.

When I run the tests which I've written to test the event/type sent from celery, I get the following traceback:

```text
======================================================================
ERROR: Test assigning tasks to hosts.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/django/test/utils.py", line 387, in inner
    return func(*args, **kwargs)
  File "/code/ews/test_controller/tests/test_tasks.py", line 67, in test_pick_host_for_task_execution
    assigned_tasks = pick_host_for_task_execution()
  File "/usr/local/lib/python3.8/site-packages/celery/local.py", line 188, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python3.8/site-packages/celery/app/task.py", line 388, in __call__
    return self.run(*args, **kwargs)
  File "/code/ews/test_controller/tasks.py", line 56, in pick_host_for_task_execution
    asyncio.run(send_assign_task(group_name=group_name, task=task, host=host))
  File "/usr/local/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/code/ews/test_controller/tasks.py", line 22, in send_assign_task
    await get_channel_layer().group_send(
  File "/usr/local/lib/python3.8/site-packages/channels_rabbitmq/core.py", line 444, in group_send
    connection = await self.carehare_connection
  File "/usr/local/lib/python3.8/site-packages/channels_rabbitmq/core.py", line 182, in carehare_connection
    raise RuntimeError(
RuntimeError: The caller tried using channels_rabbitmq on a different event loop than the one it was initialized with.
-------------------- >> begin captured logging << --------------------
asyncio: DEBUG: Using selector: EpollSelector
asyncio: DEBUG: Using selector: EpollSelector
channels_rabbitmq.core: DEBUG: group_send {'type': 'task.assign', 'task_id': '9eb2fd400a7441c69a00f7ab9bf276e7', 'host_id': '51d62ff23b4e49529b7ddf65cc218d2f'} to agent_722e36ad-d92d-42db-a6af-180ab4cb6efa
carehare._protocol: DEBUG: accept_frame: <Connection.Start object at 0x7fd1bc4ca940>
carehare._frame_writer: DEBUG: send_frame: <Connection.StartOk object at 0x7fd1bbc6cea0>
carehare._protocol: DEBUG: accept_frame: <Connection.Tune object at 0x7fd1bbc6cea0>
carehare._frame_writer: DEBUG: send_frames: [<Connection.TuneOk object at 0x7fd1bbc77a40>, <Connection.Open object at 0x7fd1bbc77b30>]
carehare._heartbeat: DEBUG: Sending heartbeats every 30.000000s...
carehare._protocol: DEBUG: accept_frame: <Connection.OpenOk object at 0x7fd1bbc6f6c0>
carehare._frame_writer: DEBUG: send_frames: [<Channel.Open object at 0x7fd1bbc74ac0>, <Exchange.Declare object at 0x7fd1bc4bfec0>, <Channel.Close object at 0x7fd1bbc7d9a0>]
carehare._protocol: DEBUG: accept_frame: <Channel.OpenOk object at 0x7fd1bbc6d080>
carehare._protocol: DEBUG: accept_frame: <Channel.CloseOk object at 0x7fd1bbc58160>
carehare._frame_writer: DEBUG: send_frames: [<Channel.Open object at 0x7fd1bbc74f40>, <Queue.Declare object at 0x7fd1bc4b4f20>, <Channel.Close object at 0x7fd1bbc86450>]
carehare._protocol: DEBUG: accept_frame: <Channel.OpenOk object at 0x7fd1bbc6f740>
carehare._protocol: DEBUG: accept_frame: <Channel.CloseOk object at 0x7fd1bbc58160>
channels_rabbitmq.core: INFO: Connected to RabbitMQ
carehare._frame_writer: DEBUG: send_frames: [<Channel.Open object at 0x7fd1bbc889c0>, <Basic.Qos object at 0x7fd1bbc868b0>, <Basic.Consume object at 0x7fd1bc4b4eb0>]
carehare._frame_writer: DEBUG: send_frames: [<Channel.Open object at 0x7fd1bbc8a8c0>, <Confirm.Select object at 0x7fd1bbc8aa00>]
carehare._frame_writer: DEBUG: send_frames: [<Basic.Publish object at 0x7fd1bc4ca700>, <pamqp.header.ContentHeader object at 0x7fd1bbc583a0>, <pamqp.body.ContentBody object at 0x7fd1bbc58400>]
carehare._protocol: DEBUG: accept_frame: <Channel.OpenOk object at 0x7fd1bbc8f740>
carehare._protocol: DEBUG: accept_frame: <Basic.QosOk object at 0x7fd1bbc58340>
carehare._protocol: DEBUG: accept_frame: <Channel.OpenOk object at 0x7fd1bbc74200>
carehare._protocol: DEBUG: accept_frame: <Basic.Ack object at 0x7fd1bbc74200>
carehare._consume_channel: ERROR: Closing consumer
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/channels_rabbitmq/reader.py", line 32, in consume_into_multi_queue_until_connection_close
    body, delivery_tag = await consumer.next_delivery()
  File "/usr/local/lib/python3.8/site-packages/carehare/_consume_channel.py", line 196, in next_delivery
    return await _next_delivery(self._queue, self.closed)
  File "/usr/local/lib/python3.8/site-packages/carehare/_consume_channel.py", line 39, in _next_delivery
    done, pending = await asyncio.wait(
  File "/usr/local/lib/python3.8/asyncio/tasks.py", line 426, in wait
    return await _wait(fs, timeout, return_when, loop)
  File "/usr/local/lib/python3.8/asyncio/tasks.py", line 534, in _wait
    await waiter
asyncio.exceptions.CancelledError
carehare._frame_writer: DEBUG: send_frame: <Channel.Close object at 0x7fd1bbc97450>
carehare._protocol: DEBUG: accept_frame: <Channel.CloseOk object at 0x7fd1bbc584c0>
asyncio: DEBUG: Using selector: EpollSelector
channels_rabbitmq.core: DEBUG: group_send {'type': 'task.assign', 'task_id': '9eb2fd400a7441c69a00f7ab9bf276e7', 'host_id': '945e15abb9994e1ca705cf4ce350d242'} to agent_722e36ad-d92d-42db-a6af-180ab4cb6efa
--------------------- >> end captured logging << ---------------------
```

I guess that the problem lies in the fact that I'm using channels_rabbitmq on a different event loop than the one it was initialized with as the logs say, but since I'm new to async programming I would really appreciate your help here. What am I doing wrong? Is it possible to send messages to the channel_layer from a celery task in order for them to be consumed by a channels websocket consumer and update the connected clients?

I'm using celery 5.

Many thanks in advance!

**Comment by adamhooper** [comment]
Hi there,

The question shouldn't be "what am I doing wrong?" -- it should be, "how in the hell would anybody expect me to figure this out?"

TL;DR I recommend against Celery. (I don't understand what value it delivers.)

TL;DR number 2: You can write a simple worker in [carehare](https://github.com/CJWorkbench/carehare). There's some (sparse) documentation in the Carehare docs about how to do that. In your Carehare-based code you'll have a RabbitMQ `connection`. With that, you can write, `await connection.publish(msgpack.packb({"type": "task.assign", "task_id": "id_placeholder", "host_id": "id_placeholder"}), routing_key=channel_id))`. (You'll have to encode `channel_id` in the message somehow, I guess. And certainly wrap your `publish()` in a try/catch, because you can expect `ChannelClosedException` every time a channel goes away after submitting a task.)

If it's a one-paragraph fix in Carehare, how can it be so hard in Celery? Because we maintainers all form a conspiracy of non-problem-solvers:

* Celery hates running `asyncio` tasks in the background, because that would break task isolation. (Task isolation is one of the main purposes of Celery.) Celery runs its _own_ AMQP connection in the background; but I don't think it lets you send your own messages on that connection, and you'd probably get undefined behavior even if you _could_ use that connection.)
* Carehare (disclaimer: I'm the maintainer) hates non-async code, because RabbitMQ's protocol is asynchronous and clients must take instruction from servers. And it hates undefined behavior, because I've tried four other RabbitMQ client libraries in production :).)
* Channels and its protocol, `asgiref`, hate to tell you "no", because it's "Pythonic" to _not_ tell you "no" when you type something weird.
* And _all of us_ hate to agree on a way to achieve the simple, common task you are trying to achieve.

(Welcome to the Python community :).)

How this plays out in code: `async_to_sync()` works in Django views and Django-Channels consumers. Outside of Django views and Django-Channels consumers, `async_to_sync()` does not work. **In Celery, `async_to_sync()` does not work.**

You can't avoid the fact that it doesn't work. You can only pick your poison. `channels_rabbitmq` tells you it doesn't work by raising a `RuntimeError`. `channels_redis` tells you it doesn't work by ... uh ... pretending to work, reconnecting every time any message is sent, and causing massive slowdowns.

At Workbench, we avoid this circle of neglect by avoiding Celery. (We use microservices, so Celery is the wrong tool for us anyway.) _You_ can break the circle of neglect by nixing Celery from your stack; nixing `channels_rabbitmq` from your stack; or nixing Channels from your stack.

You won't find good documentation for any of these routes, and yet you must choose one. I think the Carehare path is the closest to the dream; but I admit it's very hard to grok. I'm stretched thin here....

I'll close the issue because I think I've summed up: **channels_rabbitmq won't work with Celery**.

What's next for you? I recommend you try to implement a Carehare worker. If you're having trouble building a Carehare worker, please open issues there: I'll help, and we can collaborate on Carehare documentation. If your Carehare worker seems okay except for the fact that you can't send messages from Carehare to `channels_rabbitmq`, then please open issues _here_: I'll help, and we can collaborate on the documentation.

**Comment by EugenSusurrus** [comment]
Hi @adamhooper,

many thanks for taking the time to answer my question. I can imagine that you've been repeating yourself a lot regarding Celery so I really appreciate that you've taken the time to make it clear again. However I'm still confused, now I'm thinking to follow the next strategy (since "nixing Celery" from our stack is not going to be easy.)

I was thinking to:

**1. Write a `carehare` worker.** Would it then be possible to run the Carehare worker and the `celery` worker together?
**2. From the `celery` worker publish a message to the `carehare`**. Using something similar to what you've proposed:
```python
async def enqueue_task_or_raise_serious_error(task: bytes) -> None:
    channel_layer = get_channel_layer()
    carehare_connection = await channel_layer.carehare_connection

    # This next line doesn't belong here: it belongs in your app's startup....
    await carehare_connection.queue_declare("my-tasks")  # only needed once, on app startup

    await carehare_connection.publish(task, routing_key="my-tasks")
```
**3. When creating a `carehare` worker I'm thinking if it would be possible to trigger it in the [*Django `ready()` hook*](https://docs.djangoproject.com/en/3.2/ref/applications/#methods). Or otherwise how would it be possible to make it run in the `runserver` command without the need to create a new command to run the worker?
**4. From the `carehare` worker I want to then route the message to the `channels` Websocket consumer**. Hopefully that would work but it seems sketchy. What are you thoughts on that?

Many thanks again!

**Comment by adamhooper** [comment]
> **1. Write a `carehare` worker.** Would it then be possible to run the Carehare worker and the `celery` worker together?

Yes, in a separate process.

The carehare worker can be a Django app, too. So it can access your database, and it can even push tasks onto Celery queues.

Your `celery` worker would be able to execute tasks off your Celery queues, as it does now. Your Carehare worker would be able to execute tasks off your Carehare queues _and write messages to Django Channels_.

Longer-term, you may end up moving more and more tasks into the Carehare pattern, until you delete Celery from your stack entirely. Or you may dislike the Carehare approach and move everything to Celery and figure out some other way to send messages from Celery to Channels. (Maybe a synchronous RabbitMQ client that is somehow embedded within a Celery process? I never found a synchronous RabbitMQ client that can handle errors the way I want, but you might.)

> **2. From the `celery` worker publish a message to the `carehare`**.

As far as I know, this is not going to work. _Somewhere_ in your Celery worker, if you want to use carehare, you'll be forced call `async_to_sync()` or `asyncio.run()`. If it's not in your own codebase, it's in a dependency's codebase. `channels_rabbitmq` does not support this.

To reiterate: as long as Celery doesn't allow long-running asyncio tasks, the only way to send messages over an async client is to disconnect and reconnect to RabbitMQ every time you run a task. You can find libraries that do this (like `channels_redis`) ... but `channels_rabbitmq` doesn't and won't because I, not having seen your infrastructure, am already certain that it's a terrible idea in your specific case :).

> **3. When creating a `carehare` worker I'm thinking if it would be possible to trigger it in the** [*Django `ready()` hook*](https://docs.djangoproject.com/en/3.2/ref/applications/#methods). Or otherwise how would it be possible to make it run in the `runserver` command without the need to create a new command to run the worker?

Definitely create a new command.

(some random tips follow...:)

You'll want an `async def main()` that reads tasks from the queue; and your Django command should run `asyncio.run(main())`.

Create a new carehare worker per work queue. Run each carehare worker in a separate process.

You'll probably want `ASGI_THREADS=2` on the workers' environments, or even `ASGI_THREADS=1` if it works: all your Django database calls must be wrapped in `database_sync_to_async()` and you'll probably only run a single SQL query at a time, so a big thread pool is a waste of resources on the database side.

> **4. From the `carehare` worker I want to then route the message to the `channels` Websocket consumer**. Hopefully that would work but it seems sketchy. What are you thoughts on that?

Yes, that's what my line of code in my previous comment was sketching out. (I'm sure it's a bit buggy.)

I agree it looks sketchy; but it's "supported": `channels_rabbitmq` won't change in between major versions in a manner that would break your code that does this.

A `group_send()` looks like:

```
async def send_to_group(connection: carehare.Connection, group_name: str, data: Dict[str, Any]) -> None:
    channels_data = dict(__asgi_group__=group_name, **data)
    try:
        await connection.publish(
            msgpack.packb(channels_data),
            exchange_name="groups",  # magic string for channels_rabbitmq
            routing_key=group_name,
        )
    except carehare.ServerSentNack:
          logger.warning(
              "Did not deliver to all queues on group %r: a queue is at capacity",
              group_name,
          )
```

A `channel_send()` might look like:

```
async def send_to_group(connection: carehare.Connection, channel_id: str, data: Dict[str, Any]) -> None:
    channels_data = dict(__asgi_channel__=group_name, **data)
    # channel_id looks like "channels_abc123!asdfasdf".
    # RabbitMQ has one queue per server, called "channels_abc123".
    queue_name = channel_id.slice(0, channel_id.index('!'))
    try:
        await connection.publish(
            msgpack.packb(channels_data), routing_key=queue_name
        )
    except carehare.ServerSentNack:
          logger.warning(
              "Did not deliver to channel %r: queue is at capacity or web server went away",
              channel_id,
          )
```

**Comment by EugenSusurrus** [comment]
Hi @adamhooper ,

Just a quick update on the current state of my struggles :)

To get around the `async` limitations of `celery` I have 2 queues on my RabbitMQ broker, both on the `default` exchange.

1. `task_queue` - To handle `@task` messages for Celery workers
2. `mq_mesages` - To handle other messages received by my `consume_rabbitmq` management command from a `celery` task.

The RabbitMQ consumer is:

```python
  import asyncio
  import json
  import logging
  
  from channels.layers import get_channel_layer
  from django.core.management.base import BaseCommand, CommandError
  
  
  logger = logging.getLogger("django-simple")
  
  
  def get_routing_key():
      """Returns the routing key / queue name used by the `consume_rabbitmq` worker."""
      return "mq_messages"
  
  
  class Command(BaseCommand):
      help = "Connects to RabbitMQ Queue. Consumes the messages and routes them to the Django Channels Consumer"
  
      @staticmethod
      async def _handle_message(message: bytes) -> None:
          logger.info(f"RECEIVED {message}")
  
          # Process the message
          message = json.loads(message)
          message_type = message.pop("type")
          message_group_name = message.pop("group_name")
  
          # Route the message to the Channels Consumer
          await get_channel_layer().group_send(
              message_group_name, {"type": message_type, "message": message}
          )
  
      @staticmethod
      async def _consume_messages():
          channel_layer = get_channel_layer()
  
          carehare_connection = await channel_layer.carehare_connection
  
          async with carehare_connection.acking_consumer(get_routing_key()) as consumer:
              async for message in consumer:
                  await Command._handle_message(message)
  
      def handle(self, *args, **options):
          logger.info(f"Connect to RabbitMQ and consume from the queue: {get_routing_key()}")
          asyncio.run(self._consume_messages())
```

I register the `mq_messages` queue somewhere in `celery.py` like so:

```python
# ...
from kombu import Exchange, Queue

# Declare the Queue to be used in
with app.pool.acquire(block=True) as connection:
    queue = Queue(
        name="mq_messages",
        exchange=Exchange(settings.CELERY_DEFAULT_EXCHANGE),
        routing_key=get_routing_key(),
        channel=connection,
    )
    queue.declare()

# ...
```

Then from the `celery` `@task` I'm able to publish messages to the `mq_messages` like so':

```python
from celery import current_app, shared_task
from channels.layers import get_channel_layer
from django.conf import settings

def publish_message_to_broker(message, routing_key):
    with current_app.producer_pool.acquire(block=True) as producer:
        producer.publish(
            message,
            routing_key=routing_key,
            exchange=settings.CELERY_DEFAULT_EXCHANGE,
            retry=True,
        )

@shared_task
def pick_host_for_task_execution():
    from my_app.models import Host
    from my_app.management.commands.consume_rabbitmq import get_routing_key

    routing_key = get_routing_key()

    for _ in range(10):
        publish_message_to_broker(
            message={"type": f"task.assign",
            "group_name": host.agent.group_name. # dynamic group_name as the client connects
            },
            routing_key=routing_key,
        )

```

But for some reason I can't send the messages from the `consume_rabbitmq` management command to the websocket consumer.

Also when connecting, a client through websocket, it seems like I can't add it to a group, when inspecting `self.groups` with the debugger after calling `async_to_sync(self.channel_layer.group_add)(self._group_name, self.channel_name)` in the `receive` method in the `consumers.py` the value of the `self.groups` is still `[]`. I feel like I'm really close to figuring this out but at the same time I'm stuck. Any further advise would be much appreciated.

**Comment by EugenSusurrus** [comment]
Hi @adamhooper ,

many thanks for the fast response. As I've ditched the Management Command looks like the `groups` exchange is not there anymore.

```
amqp.exceptions.NotFound: Basic.publish: (404) NOT_FOUND - no exchange 'groups' in vhost '/'
```

When does this exchange get created? As the `django-channels` app is receiving websocket connections? Was `groups` previously declared by `carehare_connection`?

**Comment by adamhooper** [comment]
The `groups` exchange is created after an `await get_channel_layer().carehare_connection` succeeds.

In normal practice, this would occur when somebody calls `await get_channel_layer().send()`, `.group_add()`, `group_send()`, or `group_discard()`. You can force it by running `await get_channel_layer().carehare_connection`.

The name is configurable by the `groups_exchange` option.

**Comment by EugenSusurrus** [comment]
Hi @adamhooper,

after changing my function which should publish to the `groups` exchange to:
```python
async def connect_to_channel_layer():
    channel_layer = get_channel_layer()
    carehare_connection = await channel_layer.carehare_connection


def publish_message_to_broker(message, routing_key):
    asyncio.run(connect_to_channel_layer())

    with current_app.producer_pool.acquire(block=True) as producer:
        producer.publish(
            msgpack.packb(message),
            exchange="groups",
            content_encoding='binary',
            routing_key=routing_key,
            retry=False,
        )
```

I'm getting the good ol' error:

```
  File "/usr/local/lib/python3.8/site-packages/channels_rabbitmq/core.py", line 182, in carehare_connection
    raise RuntimeError(
RuntimeError: The caller tried using channels_rabbitmq on a different event loop than the one it was initialized with.
```

Any clue where should I call `await get_channel_layer().carehare_connection` in order to get rid of this problem? It seems like I'm getting closer but there are some sticks in my wheels :D

**Comment by adamhooper** [comment]
@EugenSusurrus I can think of two approaches -- both will work:

1. Declare the exchange in Django: await the `carehare_connection` -- maybe using a global variable (so you only await once) in your Channels consumer, upon connect. (This ought to happen automatically -- I'm tracking that at #38)
2. Declare the exchange in Celery -- as I gather you're trying now. I'm not quite sure how to do this. Perhaps `producer.maybe_declare(Exchange("groups"))`, using `from kombu.entity import Exchange`?

Definitely, 2 is the way to go. I don't quite know how to do it; but I hope to find out!

**Comment by EugenSusurrus** [comment]
Hi @adamhooper ,

I'm able to declare the exchange in the celery.py like so:

```python
# Declare the Exchange
with app.pool.acquire(block=True) as connection:
    exchange = kombu.Exchange(
        name='groups',
        type='direct',
        durable=True,
        channel=connection,
    )
    exchange.declare()
```

and I can successfully publish messages to the exchange from the celery task like so:

```python
def publish_message_to_broker(message, routing_key):
    with current_app.producer_pool.acquire(block=True) as producer:
        producer.publish(
            msgpack.packb(message),
            exchange="groups",
            content_encoding='binary',
            routing_key=routing_key,
            retry=False,
        )


@shared_task
def pick_host_for_task_execution():
    publish_message_to_broker(
            message={"type": "task.assign"},
            routing_key="test_group",
        )
```

My websocket consumers register on connection like so:

```python
class WebsocketConnectionConsumer(WebsocketConsumer):

    def connect(self):
        token_is_valid = self.scope.get("validated", None)
        if token_is_valid is not None and token_is_valid:
            # If "validated" field is present in the scope,
            # meaning "Authorization" header was provided
            # and it has the value `True`, meaning is a valid
            # `WebsocketToken` instance
            self.accept()

            # Add Agent's group_name to the channel layer
            # Somewhere here I get the connected_client_instance
            connected_client_instance = .....

            sync_to_async(self._add_agent_group_name_to_channel_layer)(agent_instance= connected_client_instance)
        else:
            self.close()

    def _add_agent_group_name_to_channel_layer(self, agent_instance):
        async_to_sync(self.channel_layer.group_add)("test_group", self.channel_name)

    def task_assign(self, event):
        """Handles the `task.assign` event received on the channel layer."""
        async_to_sync(self.send(json.dumps(
            {
                "event": "task.received",
                "payload": {
                }
            }
        )))
```

However I still can't trigger the `task_assign` event. Following the documentation on `django-channels` having `"type": "task.assign"` in the message body should do the trick, however it doesn't work. Any clues on what I\m doing wrong?

The celery `@task` definitely sends messages to the `groups` exchange since I can clearly see that in the RabbitMQ management http interface.

Thanks again!

**Comment by adamhooper** [comment]
Glad we're making progress!

Did you add `"__asgi_group__"` to the message? If you didn't, `channels_rabbitmq` will ignore it.

Also, there's a `sync_to_async()` in your `def connect()` that doesn't belong. This might prevent the group_add from doing anything. You can check on "bindings" in the RabbitMQ admin interface to see whether a group subscription is active.

**Comment by EugenSusurrus** [comment]
Hi @adamhooper , 

good news finally! Many thanks for the help!
Now I'm able to send messages from a `Celery` `@task` to a `Channels` consumer. Here is the basic version of the final code that does the trick:

In `tasks.py`:
```python
def publish_message_to_broker(message, routing_key):
    with current_app.producer_pool.acquire(block=True) as producer:
        # The channel layer needs to read an __asgi_group__ member.
        # Otherwise it will be ignored by the channels-rabbitmq
        message.update({"__asgi_group__":  routing_key})

        # `retry=False` because channels has at-most-once delivery semantics.
        # `content_encoding='binary'` because msgpack-ed message will be raw bytes.
        producer.publish(
            msgpack.packb(message),
            exchange="groups",
            content_encoding='binary',
            routing_key=routing_key,
            retry=False,
        )


@shared_task
def pick_host_for_task_execution():
    publish_message_to_broker(
            message={"type": "task.assign"},
            routing_key="test_group",
        )
```
in `consumers.py`
```python
class WebsocketConnectionConsumer(WebsocketConsumer):

    def connect(self):
        self.accept()

        # Add connected client to group.
        async_to_sync(self.channel_layer.group_add)("test_group", self.channel_name)

    def task_assign(self, event):
        """Handles the `task.assign` event received on the channel layer."""
        async_to_sync(self.send(json.dumps(
            {
                "event": "task.received",
                "payload": {
                }
            }
        )))
```

The behaviour is as follows:
The `@task` is running with a `cron` `beatschedule` of 1 minute, it publishes to the `groups` RabbitMQ exchange. If the exchange is not yet declared, the task fails, this is expected and desired behaviour.

The `groups` exchange is instantiated in this line `async_to_sync(self.channel_layer.group_add)("test_group", self.channel_name)`, you can see this clearly in RabbitMQ management, and a channel in bound to it with the routing key of `test_group`. This means that the exchange is declared when the first WS Channels client connects and it persists as long as there are connected clients because it is a `durable=False` exchange. Hence the Celery task will succeed as long as there are connected clients that have the correct `routing_key` to receive the message.

Here is also a stack overflow post that helped me, but in the end proved unnecessary since I\m not declaring the `groups` exchange myself anymore. The helpful part was in explaining how to get the current connection in celery and publish to the RabbitMQ.

https://stackoverflow.com/questions/48633494/publish-messages-to-exchanges-in-rabbitmq-with-celery

**Comment by adamhooper** [comment]
@EugenSusurrus Thank you ever so much for all your efforts. I've adjusted the `README.rst` to include your advice. It feels soooo good to have that missing piece of the puzzle solved.

**Comment by EugenSusurrus** [comment]
@adamhooper it was my pleasure. Thank you for the fast responses and for finding the time to guide me to the solution.
